File: ruby/my_app.rb
require 'active_record'
require 'awesome_print'
require 'json'
require 'puma'
require 'sinatra'

ActiveRecord::Base.logger = Logger.new(File.open('database.log', 'w'))

ActiveRecord::Base.establish_connection(
  :adapter  => 'sqlite3',
  :database => 'people.db'
)

ActiveRecord::Schema.define do
  if !ActiveRecord::Base.connection.tables.include? 'people'
    create_table :people do |table|
      table.column :name,  :string
      table.column :email, :string
      table.column :nationbuilder_id, :integer
      table.column :updates, :integer  # count # of updates
    end
  end
end

class Person < ActiveRecord::Base
  has_many :tracks

  def self.synchronize(name, email, id)
    person = Person.find_or_create_by(nationbuilder_id: id)
    person.name = name
    person.email = email
    person.updates = person.updates.to_i + 1

    person.save
  end
end

post '/update_person' do
  person = JSON.parse(request.body.read)
  full_name = person['first_name'] + ' ' + person['last_name']
  Person.synchronize(full_name, person['email'], person['id'])
end

get '/list_people' do
  rows = ""
  Person.find_each do |p|
    rows << <<-ROWS
    <tr>
      <td>#{p.id}</td>
      <td>#{p.name}</td>
      <td>#{p.email}</td>
      <td>#{p.updates}</td>
    </tr>
    ROWS
  end

  <<-TABLE
  <h4>People</h4>
  <table style="width:300px">
  <tr>
    <td>ID</td>
    <td>Name</td>
    <td>Email</td>
    <td>Updates</td>
  </tr>
  #{rows}
  </table>
  TABLE
end

File: go/my_app.go
package main

import (
	_ "code.google.com/p/go-sqlite/go1/sqlite3"
	"database/sql"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"log"
	"net/http"
	"os"
	"runtime"
	"strings"
)

func httpPostHandler(w http.ResponseWriter, r *http.Request) {
	if r.Method == "POST" {
		body, err := ioutil.ReadAll(r.Body)
		if err != nil {
			return
		}

		var person_map map[string]interface{}

		if err := json.Unmarshal(body, &person_map); err != nil {
			panic(err)
		}

		database := Open()
		defer database.Close()

		// f := person_map["id"].(float64)
		fullname := fmt.Sprintf("%s %s", person_map["first_name"].(string), person_map["last_name"].(string))

		p := Person{
			id:      int(person_map["id"].(float64)),
			name:    fullname,
			email:   person_map["email"].(string),
			updates: 0}

		database.Insert(p)

		fmt.Printf("Person updated: %s\n", string(fullname))
	}
}

func httpGetHandler(w http.ResponseWriter, r *http.Request) {
	if r.Method == "GET" {
		database := Open()

		defer database.Close()
		database.ListPeople(w)
	}
}

const PORT = ":4567"

func main() {
	http.HandleFunc("/person", httpPostHandler)
	http.HandleFunc("/people", httpGetHandler)

	error := http.ListenAndServe(PORT, nil)
	if error != nil {
		panic(error)
	}
}

func init() {
	// Use all CPUs
	runtime.GOMAXPROCS(runtime.NumCPU())
}


// database specific code below here

type Person struct {
	id      int
	name    string
	email   string
	updates int
}

type Database struct {
	database sql.DB
}

func Open() Database {
	database_filename := "people.db"

	db, err := sql.Open("sqlite3", database_filename)
	if err != nil {
		fmt.Println("Unable to open the database: %s", err)
		os.Exit(1)
	}

	db.Exec(`CREATE TABLE people(id INTEGER PRIMARY KEY ON CONFLICT REPLACE, name VARCHAR(20),
        email TEXT, updates INTEGER DEFAULT 1);`)

	return Database{database: *db}
}

func (db Database) Insert(p Person) {
	query := "SELECT updates FROM people where id = ?"

	rows, err := db.database.Query(query, p.id)
	if err != nil {
		log.Fatal(err)
	}

	var updates int

	defer rows.Close()
	for rows.Next() {
		rows.Scan(&updates)
	}

	updates++

	fmt.Sprintln("updates => %d", updates)

	insert := "INSERT INTO people(id, name, email, updates) VALUES(?, ?, ?, ?);"

	_, err = db.database.Exec(insert, p.id, p.name, p.email, updates)
	if err != nil {
		log.Fatal(err)
	}
}

func (db Database) ListPeople(w http.ResponseWriter) {
  rows, err := db.database.Query("SELECT id, name, email, updates FROM people;")
  if err != nil {
    log.Fatal(err)
  }

  fmt.Fprintf(w, "%-5s %-15s %-20s %-5s\n", "ID", "Name", "Email", "Updates")
  fmt.Printf(strings.Repeat("-", 66) + "\n")

  for rows.Next() {
    var p Person
    rows.Scan(&p.id, &p.name, &p.email, &p.updates)
    fmt.Fprintf(w, "%-5d %-15s %-20s %-5d\n", p.id, p.name, p.email, p.updates)
  }
}

func (db Database) Close() {
	db.database.Close()
}
